File: apps/server/providers/src/claude/oauth.ts
/**
 * Claude OAuth: PKCE authorization URL, JSON token exchange, credential
 * storage. Mirrors pi's `auth/oauth/anthropic.ts` and console's
 * `codex/oauth.ts` structure.
 */
import * as crypto from "node:crypto";
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import {
  CLAUDE_AUTHORIZE_URL,
  CLAUDE_CLIENT_ID,
  CLAUDE_REFRESH_SKEW_MS,
  CLAUDE_SCOPE,
  CLAUDE_TOKEN_URL,
  claudeRedirectUri,
} from "./constants.js";

export interface ClaudeOAuthCredential {
  access_token: string;
  refresh_token: string;
  expiresAt: number;
  email?: string;
}

export interface ParsedClaudeCredential {
  accessToken: string;
  refreshToken: string;
  expiresAtMs: number;
  email?: string;
}

interface TokenResponse {
  access_token?: string;
  refresh_token?: string;
  expires_in?: number;
}

function credentialPath(): string {
  return process.env.CLAUDE_CREDENTIALS_PATH ?? path.join(os.homedir(), ".console", "claude-creds.json");
}

export function generateClaudePkce(): { verifier: string; challenge: string } {
  const verifier = crypto.randomBytes(32).toString("base64url");
  const challenge = new Bun.CryptoHasher("sha256").update(verifier).digest("base64url");
  return { verifier, challenge };
}

export function createClaudeAuthorizationUrl(args: {
  state: string;
  verifierChallenge: string;
}): { authUrl: string; redirectUri: string } {
  const redirectUri = claudeRedirectUri();
  const url = new URL(CLAUDE_AUTHORIZE_URL);
  url.search = new URLSearchParams({
    code: "true",
    client_id: CLAUDE_CLIENT_ID,
    response_type: "code",
    redirect_uri: redirectUri,
    scope: CLAUDE_SCOPE,
    code_challenge: args.verifierChallenge,
    code_challenge_method: "S256",
    state: args.state,
  }).toString();
  return { authUrl: url.toString(), redirectUri };
}

function tokenError(status: number, body: string): Error {
  let detail = body.trim();
  try {
    const parsed = JSON.parse(detail) as { error?: unknown; error_description?: unknown; message?: unknown };
    detail = String(parsed.error_description ?? parsed.error ?? parsed.message ?? detail);
  } catch {
    // Keep the raw response when it is not JSON.
  }
  return new Error(`Claude OAuth request failed (${status}): ${detail || "unknown error"}`);
}

async function postToken(body: Record<string, string>): Promise<TokenResponse> {
  const response = await fetch(CLAUDE_TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(15_000),
  });
  if (!response.ok) throw tokenError(response.status, await response.text().catch(() => ""));
  return (await response.json()) as TokenResponse;
}

/**
 * Best-effort account identity lookup via the Claude Code bootstrap
 * endpoint (mirrors oh-my-pi's `anthropic-identity` hook). Returns the
 * account email when available, undefined otherwise — never throws.
 */
async function fetchBootstrapEmail(accessToken: string): Promise<string | undefined> {
  try {
    const response = await fetch(
      "https://api.anthropic.com/api/claude_cli/bootstrap?entrypoint=cli&model=claude-sonnet-4-6",
      {
        method: "GET",
        headers: {
          Accept: "application/json, text/plain, */*",
          Authorization: `Bearer ${accessToken}`,
          "Content-Type": "application/json",
          "anthropic-beta": "oauth-2025-04-20",
        },
        signal: AbortSignal.timeout(15_000),
      },
    );
    if (!response.ok) return undefined;
    const data = (await response.json()) as {
      oauth_account?: { account_email?: unknown };
    };
    const email = data.oauth_account?.account_email;
    return typeof email === "string" && email.length > 0 ? email : undefined;
  } catch {
    return undefined;
  }
}

export async function exchangeClaudeCode(
  code: string,
  state: string,
  verifier: string,
  redirectUri: string,
): Promise<ClaudeOAuthCredential> {
  const data = await postToken({
    grant_type: "authorization_code",
    client_id: CLAUDE_CLIENT_ID,
    code,
    state,
    redirect_uri: redirectUri,
    code_verifier: verifier,
  });
  if (!data.access_token || !data.refresh_token || typeof data.expires_in !== "number") {
    throw new Error("Claude OAuth response is missing required token fields.");
  }
  return {
    access_token: data.access_token,
    refresh_token: data.refresh_token,
    expiresAt: Date.now() + data.expires_in * 1000,
    email: await fetchBootstrapEmail(data.access_token),
  };
}

async function refreshClaudeToken(credential: ParsedClaudeCredential): Promise<ParsedClaudeCredential> {
  const data = await postToken({
    grant_type: "refresh_token",
    client_id: CLAUDE_CLIENT_ID,
    refresh_token: credential.refreshToken,
  });
  if (!data.access_token || !data.refresh_token || typeof data.expires_in !== "number") {
    throw new Error("Claude token refresh response is missing required fields.");
  }
  const updated: ClaudeOAuthCredential = {
    access_token: data.access_token,
    refresh_token: data.refresh_token,
    expiresAt: Date.now() + data.expires_in * 1000,
    email: credential.email ?? (await fetchBootstrapEmail(data.access_token)),
  };
  await saveClaudeCredential(updated);
  return parseClaudeCredential(updated);
}

export function parseClaudeCredential(raw: ClaudeOAuthCredential): ParsedClaudeCredential {
  if (!raw.access_token || !raw.refresh_token) {
    throw new Error("Invalid Claude credential. Please login again.");
  }
  return {
    accessToken: raw.access_token,
    refreshToken: raw.refresh_token,
    expiresAtMs: raw.expiresAt,
    email: raw.email,
  };
}

export async function loadClaudeCredential(): Promise<ParsedClaudeCredential> {
  const envToken = process.env.CLAUDE_OAUTH_TOKEN ?? process.env.ANTHROPIC_OAUTH_TOKEN;
  if (envToken) {
    return {
      accessToken: envToken,
      refreshToken: "",
      expiresAtMs: Number.MAX_SAFE_INTEGER,
      email: undefined,
    };
  }
  const raw = JSON.parse(await fs.readFile(credentialPath(), "utf8")) as ClaudeOAuthCredential;
  return parseClaudeCredential(raw);
}

export async function saveClaudeCredential(credential: ClaudeOAuthCredential): Promise<void> {
  const filePath = credentialPath();
  await fs.mkdir(path.dirname(filePath), { recursive: true });
  await fs.writeFile(filePath, JSON.stringify(credential, null, 2), "utf8");
}

export async function claudeCredentialExists(): Promise<boolean> {
  if (process.env.CLAUDE_OAUTH_TOKEN ?? process.env.ANTHROPIC_OAUTH_TOKEN) return true;
  try {
    await fs.access(credentialPath());
    return true;
  } catch {
    return false;
  }
}

export async function refreshClaudeIfNeeded(credential: ParsedClaudeCredential): Promise<ParsedClaudeCredential> {
  if (!credential.refreshToken || Date.now() + CLAUDE_REFRESH_SKEW_MS < credential.expiresAtMs) return credential;
  return refreshClaudeToken(credential);
}

File: apps/server-go/internal/providers/claude/oauth.go
// Claude OAuth: PKCE authorization URL, JSON token exchange, credential
// storage. Port of apps/server/providers/src/claude/oauth.ts.
package claude

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"time"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/providers/shared"
)

// OAuthCredential is the stored credential file shape.
type OAuthCredential struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresAt    int64  `json:"expiresAt"`
	Email        string `json:"email,omitempty"`
}

// ParsedCredential is the validated in-memory credential.
type ParsedCredential struct {
	AccessToken  string
	RefreshToken string
	ExpiresAtMs  int64
	Email        string
}

type tokenResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresIn    *int64 `json:"expires_in"`
}

// CredentialPath mirrors oauth.ts: CLAUDE_CREDENTIALS_PATH override,
// otherwise ~/.console/claude-creds.json.
func CredentialPath() string {
	return shared.CredentialPath("CLAUDE_CREDENTIALS_PATH", "claude-creds.json")
}

func GeneratePKCE() (verifier, challenge string, err error) {
	return shared.GeneratePKCE()
}

// AuthorizationURL builds the Claude Pro/Max PKCE authorize URL.
func AuthorizationURL(state, verifierChallenge string) (authURL, redirectURI string) {
	redirectURI = RedirectURI()
	u, _ := url.Parse(AuthorizeURL)
	q := url.Values{
		"code":                  {"true"},
		"client_id":             {ClientID},
		"response_type":         {"code"},
		"redirect_uri":          {redirectURI},
		"scope":                 {Scope},
		"code_challenge":        {verifierChallenge},
		"code_challenge_method": {"S256"},
		"state":                 {state},
	}
	u.RawQuery = q.Encode()
	return u.String(), redirectURI
}

func postToken(client *http.Client, body map[string]string) (tokenResponse, error) {
	if client == nil {
		client = &http.Client{Timeout: 15 * time.Second}
	}
	rawBody, err := json.Marshal(body)
	if err != nil {
		return tokenResponse{}, err
	}
	req, err := http.NewRequest("POST", TokenURL, bytes.NewReader(rawBody))
	if err != nil {
		return tokenResponse{}, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return tokenResponse{}, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return tokenResponse{}, shared.TokenError("Claude", resp.StatusCode, string(raw))
	}
	var data tokenResponse
	if err := json.Unmarshal(raw, &data); err != nil {
		return tokenResponse{}, fmt.Errorf("decode Claude token response: %w", err)
	}
	return data, nil
}

// fetchBootstrapEmail is a best-effort account identity lookup via the
// Claude Code bootstrap endpoint. It never throws (mirrors oauth.ts).
func fetchBootstrapEmail(client *http.Client, accessToken string) string {
	if client == nil {
		client = &http.Client{Timeout: 15 * time.Second}
	}
	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, "GET",
		"https://api.anthropic.com/api/claude_cli/bootstrap?entrypoint=cli&model=claude-sonnet-4-6", nil)
	if err != nil {
		return ""
	}
	req.Header.Set("Accept", "application/json, text/plain, */*")
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("anthropic-beta", "oauth-2025-04-20")
	resp, err := client.Do(req)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return ""
	}
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	var payload struct {
		OAuthAccount *struct {
			AccountEmail string `json:"account_email"`
		} `json:"oauth_account"`
	}
	if err := json.Unmarshal(raw, &payload); err != nil {
		return ""
	}
	if payload.OAuthAccount == nil || payload.OAuthAccount.AccountEmail == "" {
		return ""
	}
	return payload.OAuthAccount.AccountEmail
}

// ExchangeCode exchanges an authorization code for stored credentials.
func ExchangeCode(client *http.Client, code, state, verifier, redirectURI string) (OAuthCredential, error) {
	data, err := postToken(client, map[string]string{
		"grant_type":    "authorization_code",
		"client_id":     ClientID,
		"code":          code,
		"state":         state,
		"redirect_uri":  redirectURI,
		"code_verifier": verifier,
	})
	if err != nil {
		return OAuthCredential{}, err
	}
	if data.AccessToken == "" || data.RefreshToken == "" || data.ExpiresIn == nil {
		return OAuthCredential{}, fmt.Errorf("Claude OAuth response is missing required token fields.")
	}
	return OAuthCredential{
		AccessToken:  data.AccessToken,
		RefreshToken: data.RefreshToken,
		ExpiresAt:    time.Now().UnixMilli() + *data.ExpiresIn*1000,
		Email:        fetchBootstrapEmail(client, data.AccessToken),
	}, nil
}

func refreshToken(client *http.Client, cred ParsedCredential) (ParsedCredential, error) {
	data, err := postToken(client, map[string]string{
		"grant_type":    "refresh_token",
		"client_id":     ClientID,
		"refresh_token": cred.RefreshToken,
	})
	if err != nil {
		return ParsedCredential{}, err
	}
	if data.AccessToken == "" || data.RefreshToken == "" || data.ExpiresIn == nil {
		return ParsedCredential{}, fmt.Errorf("Claude token refresh response is missing required fields.")
	}
	email := cred.Email
	if email == "" {
		email = fetchBootstrapEmail(client, data.AccessToken)
	}
	updated := OAuthCredential{
		AccessToken:  data.AccessToken,
		RefreshToken: data.RefreshToken,
		ExpiresAt:    time.Now().UnixMilli() + *data.ExpiresIn*1000,
		Email:        email,
	}
	if err := SaveCredential(updated); err != nil {
		return ParsedCredential{}, err
	}
	return ParseCredential(updated)
}

// ParseCredential validates a stored credential.
func ParseCredential(raw OAuthCredential) (ParsedCredential, error) {
	if raw.AccessToken == "" || raw.RefreshToken == "" {
		return ParsedCredential{}, fmt.Errorf("Invalid Claude credential. Please login again.")
	}
	return ParsedCredential{
		AccessToken:  raw.AccessToken,
		RefreshToken: raw.RefreshToken,
		ExpiresAtMs:  raw.ExpiresAt,
		Email:        raw.Email,
	}, nil
}

// LoadCredential reads CLAUDE_OAUTH_TOKEN/ANTHROPIC_OAUTH_TOKEN or the
// credential file. Env tokens never expire (parity with TS MAX_SAFE_INTEGER).
func LoadCredential() (ParsedCredential, error) {
	envToken := os.Getenv("CLAUDE_OAUTH_TOKEN")
	if envToken == "" {
		envToken = os.Getenv("ANTHROPIC_OAUTH_TOKEN")
	}
	if envToken != "" {
		return ParsedCredential{
			AccessToken:  envToken,
			RefreshToken: "",
			ExpiresAtMs:  1<<62 - 1,
			Email:        "",
		}, nil
	}
	raw, err := os.ReadFile(CredentialPath())
	if err != nil {
		return ParsedCredential{}, err
	}
	dec := json.NewDecoder(bytes.NewReader(bytes.TrimSpace(raw)))
	var cred OAuthCredential
	if err := dec.Decode(&cred); err != nil {
		return ParsedCredential{}, err
	}
	return ParseCredential(cred)
}

// SaveCredential writes the credential file (0600 dir 0700).
func SaveCredential(cred OAuthCredential) error {
	return shared.SaveCredentialFile(CredentialPath(), cred)
}

// CredentialExists reports whether a login exists (env token or file).
func CredentialExists() bool {
	if os.Getenv("CLAUDE_OAUTH_TOKEN") != "" || os.Getenv("ANTHROPIC_OAUTH_TOKEN") != "" {
		return true
	}
	return shared.CredentialFileExists(CredentialPath())
}

// RefreshIfNeeded refreshes an expiring credential (5-minute skew).
func RefreshIfNeeded(client *http.Client, cred ParsedCredential) (ParsedCredential, error) {
	if cred.RefreshToken == "" || time.Now().UnixMilli()+RefreshSkewMs < cred.ExpiresAtMs {
		return cred, nil
	}
	return refreshToken(client, cred)
}
